fix: refuse creating new DAGs with insecure hash functions (SHA1)

finkbeca opened the pull request:
Closing #8703, added an optional forcing flag for `ipfs add` to discourage use of broken hashes like SHA-1 while still allowing it, given you pass in an optional forcing flag. Tested this manually due to the small scope of the changes.

```console
$ ipfs add --hash=sha1 file
Error: selected hash function is no longer secure; use --hash=sha2-256 or pass --force.
```

```console
$ ipfs add --hash=sha1 --force file
Added ...
```

guseggert commented:
Thanks for this!

A more specific flag would make it more obvious that a less secure hash function is being used, like `--allow-insecure-hash-function` or something like that. Also can you add a sharness test for this?  Should be really easy.

finkbeca commented:
Thanks for the feedback. I fixed suggested the edits changing the flag name to `--allow-insecure-hash-function` and added a sharness test for the new command.  Hope that cover the change requests. :)

Jorropo commented on test/sharness/t0047-add-forcehash.sh:
```diff
@@ -0,0 +1,47 @@
+    HASH="QmVr26fY1tKyspEJBniVhqxQeEjhF78XerGiqWAwraVLQH" &&
+    echo "added $HASH hello.txt" >expected &&
+    test_cmp expected actual
+  '
```
You should add a test that tries to add a file with an unsecure function and check that it *fails*.

Jorropo commented on core/commands/add.go:
```diff
@@ -181,11 +185,15 @@ only-hash, and progress/status related flags) will change the final hash.
 		inlineLimit, _ := req.Options[inlineLimitOptionName].(int)
 
+		forceHash, _ := req.Options[forceHashOptionName].(bool)
 		hashFunCode, ok := mh.Names[strings.ToLower(hashFunStr)]
+		if !forceHash && strings.ToLower(hashFunStr) == "sha1" {
```
```suggestion
		loweredHash := strings.ToLower(hashFunStr)
		hashFunCode, ok := mh.Names[loweredHash]
		if !forceHash && loweredHash == "sha1" {
```

Jorropo reviewed (changes requested):
Thanks, this is definitely a feature we want to merge, however a few things needs some tweaking first. :slightly_smiling_face:

finkbeca commented:
Fixed the syntax change @Jorropo , also added failure tests specifically for SHA1. I'm aware that the Patch only works for SHA1 at this time, I can change the wording of the error if you think that would be better. My initial thought that generalizing the error statement will simplify future extensions.

lidel reviewed (changes requested):
Thank you, @finkbeca :heart: 

Would you have time to add the same flag to `dag put` and `block put` – see  ipfs/go-ipfs#8696 for prior art of (`--allow-big-block`)?  If not, let us know, someone else will push this over the finish line.

Jorropo reviewed (changes requested):
Sorry to ask for more but this does not work.
```
ipfs add --hash=md5 --allow-insecure-hashes randomFileHere
```
Doesn't work.

I know this is a more complicated task, but I would like to avoid yet an other half working feature.
I don't know what other maintainers think of this.

lidel commented:
@Jorropo Sounds sensible. Do you have more things we should block?

@finkbeca Mind replacing static "sha1" with a check against a list of insecure hashes? 
That way, we can add them to some constant over time.

Jorropo commented:
> Mind replacing static "sha1" with a check against a list of insecure hashes ["sha1", "md5"]?
That way, we can add more as time goes by.

We already have that, it is <https://github.com/ipfs/go-verifcid>.

> Do you have more things we should block?

- [ ] Refactor go-verify CID consumers, to work on the command level (instead of deeper in the code like in add's internal) so we can disable it more easly. (what I am already asking for)
- [ ] Remove SHA1 from go-verify CID's safe list. (or only allows the "safe" cryptoanalysis fix, see https://shattered.io/)

Jorropo commented:
> I don't see a particularly good reason to bother with doing go-verifcid plumbing everywhere to allow sets of insecure hash functions. Note: you likely wouldn't want it to be all or nothing... allowing SHA-1 doesn't necessarily mean allowing a 1 bit SHA-2 hash or MD5.

I'm saying this is what we *currently* more less have.
I want to scrape that, and only do the checking in commands handlers (where adding `--allow-insecure-hash` is trivial to implement).

> Nothing at the library or protocol level stops you from using broken hash functions like MD4, but I don't see a reason why go-ipfs should support using it.
```
$ rgrep -I "verifcid" vendor/ | cut -f1 -d':' | while read f; do dirname ${f}; done | uniq | grep -v verifcid | grep -v "^vendor$"
vendor/github.com/ipfs/go-ipfs-provider/simple
vendor/github.com/ipfs/go-ipfs-provider/batched
vendor/github.com/ipfs/go-blockservice
```

aschmahmann commented:
> I want to scrape that, and only do the checking in commands handlers (where adding --allow-insecure-hash is trivial to implement).

This seems unlikely to work. For example `ipfs cat bafysomesha2` internally has a pointer to an MD4 hash -> game over. You still need this pretty low down the stack.

> Nothing at the library or protocol level stops you from using broken hash functions like MD4, but I don't see a reason why go-ipfs should support using it.

Ok fair enough there are a few library pieces where you'd have to mock out the interfaces or do a fork to remove the checks. IIRC those structs aren't explicitly used rather than their interfaces so not too bad to deal with. However, if there's actual demand for it it's probably fine to add some configurability.

Jorropo commented:
> This seems unlikely to work. For example `ipfs cat bafysomesha2` internally has a pointer to an MD4 hash -> game over. You still need this pretty low down the stack.

Ok fair enough too.

We can put that value in the context (create a `allow-insecure-hash` bool key), and let verifcid take a context and use that to know if we allows it or not. (we can't use globals obviously since this parameter is request dependent)

Jorropo commented:
Sorry but we are not intrested in this fix,
It's not actually doing what it claims, I can't use it to add files with MD4.

**But mainly.**
it only supports `ipfs add --hash=sha1`
Which is not what is important.

**The danger is in `ipfs get`** (and gateway, ipfs cat, ... whatever that read data), because some malicious attacker can create a SHA256 block linking a bad SHA1 hash and send you whatever data they want instead.

We will follow the same route as git, using the sha1 fix to make the function "safe" again. (the only reason sha1 is still there is for git anyway so using the same fix as it make sense).

The context for this is here: https://github.com/ipfs/go-ipfs/issues/8703#issuecomment-1155758989
